File: forms/organ-donation-assessment/full-stack-with-loco-tera-htmx-alpine/tests/engine/donation_grader_test.rs
use chrono::Datelike;
use organ_donation_assessment_tera_crate::engine::donation_grader::{evaluate_rules, grade, run_rules};
use organ_donation_assessment_tera_crate::engine::donation_rules::all_rules;
use organ_donation_assessment_tera_crate::engine::types::*;

/// Baseline "ideal donor" case: deceased donor with no penalties.
fn baseline_ideal_donor() -> AssessmentData {
    AssessmentData {
        demographics: Demographics {
            first_name: "Jane".to_string(),
            last_name: "Doe".to_string(),
            date_of_birth: "1990-01-01".to_string(),
            sex: "female".to_string(),
            weight: Some(70.0),
            height: Some(170.0),
            bmi: Some(24.2),
            ethnicity: "White British".to_string(),
        },
        donor_type_registration: DonorTypeRegistration {
            donor_type: "deceased".to_string(),
            ..DonorTypeRegistration::default()
        },
        ..AssessmentData::default()
    }
}

#[test]
fn empty_case_grades_as_suitable_low() {
    let data = AssessmentData::default();
    let result = grade(&data);
    assert_eq!(result.suggested_eligibility, "suitable");
    assert_eq!(result.risk_level, "low");
}

#[test]
fn baseline_ideal_donor_is_suitable_low() {
    let data = baseline_ideal_donor();
    let result = grade(&data);
    assert_eq!(result.suggested_eligibility, "suitable");
    assert_eq!(result.risk_level, "low");
}

#[test]
fn hiv_positive_yields_unsuitable_critical() {
    let mut data = baseline_ideal_donor();
    data.infectious_disease_screening.hiv_status = "positive".to_string();
    let result = grade(&data);
    assert_eq!(result.suggested_eligibility, "unsuitable");
    assert_eq!(result.risk_level, "critical");
    assert!(result.fired_rules.iter().any(|r| r.id == "ID-001"));
}

#[test]
fn abo_incompatibility_yields_unsuitable_critical() {
    let mut data = baseline_ideal_donor();
    data.immunological_assessment.abo_compatibility = "incompatible".to_string();
    let result = grade(&data);
    assert_eq!(result.suggested_eligibility, "unsuitable");
    assert_eq!(result.risk_level, "critical");
    assert!(result.fired_rules.iter().any(|r| r.id == "IM-001"));
}

#[test]
fn diabetes_yields_conditionally_suitable_moderate() {
    let mut data = baseline_ideal_donor();
    data.medical_history.has_diabetes = "yes".to_string();
    let result = grade(&data);
    assert_eq!(result.suggested_eligibility, "conditionally-suitable");
    assert_eq!(result.risk_level, "moderate");
    assert!(result.fired_rules.iter().any(|r| r.id == "MH-006"));
}

#[test]
fn egfr_50_yields_conditionally_suitable_high() {
    let mut data = baseline_ideal_donor();
    data.organ_function.egfr = Some(50.0);
    let result = grade(&data);
    assert_eq!(result.suggested_eligibility, "conditionally-suitable");
    assert_eq!(result.risk_level, "high");
    assert!(result.fired_rules.iter().any(|r| r.id == "OF-002"));
}

#[test]
fn asa_grade_i_is_positive_marker_does_not_lower_grade() {
    let mut data = baseline_ideal_donor();
    data.surgical_assessment.asa_grade = "I".to_string();
    let result = grade(&data);
    // SU-001 (ASA-I) is excluded from penalty counting.
    assert_eq!(result.suggested_eligibility, "suitable");
    assert_eq!(result.risk_level, "low");
    assert!(result.fired_rules.iter().any(|r| r.id == "SU-001"));
}

#[test]
fn living_minor_yields_unsuitable_critical() {
    let mut data = baseline_ideal_donor();
    data.donor_type_registration.donor_type = "living".to_string();
    let today = chrono::Utc::now().date_naive();
    let dob = format!("{}-01-01", today.year() - 10);
    data.demographics.date_of_birth = dob;
    let result = grade(&data);
    assert_eq!(result.suggested_eligibility, "unsuitable");
    assert!(result.fired_rules.iter().any(|r| r.id == "DM-003"));
}

#[test]
fn assessor_decision_overrides_eligibility_but_keeps_risk_level() {
    let mut data = baseline_ideal_donor();
    data.medical_history.has_diabetes = "yes".to_string();
    data.eligibility_allocation.eligibility_decision = "unsuitable".to_string();
    let result = grade(&data);
    // Engine-suggested is conditionally-suitable, but assessor overrides.
    assert_eq!(result.suggested_eligibility, "conditionally-suitable");
    assert_eq!(result.eligibility, "unsuitable");
    // Risk level reflects engine grading.
    assert_eq!(result.risk_level, "moderate");
}

#[test]
fn rule_ids_are_unique() {
    let rules = all_rules();
    let mut ids: Vec<&str> = rules.iter().map(|r| r.id).collect();
    ids.sort();
    let len_before = ids.len();
    ids.dedup();
    assert_eq!(ids.len(), len_before, "rule IDs must be unique");
}

#[test]
fn evaluate_and_run_rules_agree() {
    let data = baseline_ideal_donor();
    let a = evaluate_rules(&data);
    let b = run_rules(&data);
    assert_eq!(a.len(), b.len());
}

File: forms/organ-donation-assessment/front-end-with-svelte/src/lib/engine/donation-grader.test.ts
import { describe, it, expect } from 'vitest';
import { gradeDonor } from './donation-grader';
import { detectAdditionalFlags } from './flagged-issues';
import { donationRules } from './donation-rules';
import type { AssessmentData } from './types';

/**
 * A blank assessment with all fields at their unanswered defaults. Defined
 * locally (rather than importing the store's `createDefaultAssessment`) so the
 * engine tests stay free of `$app` and Svelte-runes module dependencies.
 */
function blankAssessment(): AssessmentData {
	return {
		demographics: { firstName: '', lastName: '', dateOfBirth: '', sex: '', weight: null, height: null, bmi: null, ethnicity: '' },
		donorTypeRegistration: { donorType: '', registeredOnDonorRegister: '', registryName: '', registrationDate: '', recipientRelationship: '', recipientName: '', previousDonation: '', previousDonationDetails: '', intendedOrgans: '' },
		medicalHistory: { hasMalignancy: '', malignancyDetails: '', hasCnsMalignancy: '', hasAutoimmuneDisease: '', autoimmuneDetails: '', hasDiabetes: '', diabetesDetails: '', hasHypertension: '', hypertensionDetails: '', hasCardiovascularDisease: '', cardiovascularDetails: '', hasActiveInfection: '', activeInfectionDetails: '', hasUncontrolledSepsis: '', hasCjdRisk: '', cjdDetails: '', ivDrugUseHistory: '', currentMedications: '', previousSurgery: '', surgeryDetails: '' },
		organFunction: { creatinine: null, egfr: null, kidneyImaging: '', kidneyNotes: '', alt: null, ast: null, bilirubin: null, liverImaging: '', liverNotes: '', ejectionFraction: null, echocardiogram: '', cardiacNotes: '', pao2Fio2Ratio: null, chestImaging: '', pulmonaryNotes: '', fastingGlucose: null, hba1c: null, pancreaticNotes: '', severeOrganFailure: '', severeOrganFailureDetails: '' },
		infectiousDiseaseScreening: { hivStatus: '', hbsAg: '', hbcAb: '', hcvAb: '', htlvStatus: '', cmvStatus: '', ebvStatus: '', syphilisScreen: '', toxoplasmaStatus: '', tuberculosisScreen: '', recentTravel: '', travelDetails: '', recentInfection: '', infectionDetails: '' },
		immunologicalAssessment: { donorBloodGroup: '', recipientBloodGroup: '', aboCompatibility: '', hlaA: '', hlaB: '', hlaC: '', hlaDr: '', hlaDq: '', hlaDp: '', hlaMatchLevel: '', crossmatchResult: '', pra: null, donorSpecificAntibodies: '', dsaDetails: '' },
		surgicalAssessment: { asaGrade: '', previousAnaesthetic: '', anaestheticComplications: '', complicationDetails: '', mallampatiScore: '', airwayConcerns: '', airwayDetails: '', surgicalFitness: '', surgicalFitnessNotes: '', plannedProcedure: '', smokingStatus: '', alcoholUse: '' },
		psychologicalAssessment: { mentalCapacityConfirmed: '', understandsProcedure: '', understandsRisks: '', voluntaryDecision: '', coercionConcerns: '', coercionDetails: '', ambivalence: '', ambivalenceDetails: '', anxietyAboutProcedure: '', previousPsychologicalIssues: '', psychologicalIssueDetails: '', supportNetwork: '', willingToProceed: '' },
		ethicalLegalRequirements: { htaAct2004Compliant: '', independentAssessorReview: '', independentAssessorName: '', independentAssessorDate: '', informedConsentGiven: '', consentFormSigned: '', consentDate: '', witnessName: '', witnessRole: '', informationLeafletProvided: '', questionsAnswered: '', financialRewardCheck: '', ethicsCommitteeApproval: '', ethicsApprovalReference: '' },
		eligibilityAllocation: { eligibilityDecision: '', eligibilityConditions: '', deferralReason: '', deferralDuration: '', allocatedOrgans: '', intendedRecipientCentre: '', assessorName: '', assessorRole: '', assessmentDate: '', additionalNotes: '' }
	};
}

/** An ideal living donor: ASA I, all screens negative, consent complete. */
function createIdealLivingDonor(): AssessmentData {
	const d = blankAssessment();
	d.demographics = { ...d.demographics, firstName: 'Alex', lastName: 'Ng', dateOfBirth: '1985-03-10', sex: 'male', weight: 75, height: 180, bmi: 23.1 };
	d.donorTypeRegistration = { ...d.donorTypeRegistration, donorType: 'living', registeredOnDonorRegister: 'yes' };
	d.organFunction = { ...d.organFunction, egfr: 100, alt: 25, ejectionFraction: 60 };
	d.infectiousDiseaseScreening = { ...d.infectiousDiseaseScreening, hivStatus: 'negative', hbsAg: 'negative', hcvAb: 'negative' };
	d.immunologicalAssessment = { ...d.immunologicalAssessment, aboCompatibility: 'compatible', crossmatchResult: 'compatible' };
	d.surgicalAssessment = { ...d.surgicalAssessment, asaGrade: 'I' };
	d.psychologicalAssessment = { ...d.psychologicalAssessment, mentalCapacityConfirmed: 'yes', understandsProcedure: 'yes', understandsRisks: 'yes', voluntaryDecision: 'yes', coercionConcerns: 'no', ambivalence: 'no', willingToProceed: 'yes', supportNetwork: 'yes' };
	d.ethicalLegalRequirements = { ...d.ethicalLegalRequirements, htaAct2004Compliant: 'yes', independentAssessorReview: 'yes', informedConsentGiven: 'yes', consentFormSigned: 'yes', financialRewardCheck: 'yes', questionsAnswered: 'yes', ethicsCommitteeApproval: 'yes' };
	return d;
}

describe('Organ Donation Grading Engine', () => {
	it('classifies an ideal donor as suitable / low risk with no penalty rules', () => {
		const result = gradeDonor(createIdealLivingDonor());
		expect(result.suggestedEligibility).toBe('suitable');
		expect(result.eligibility).toBe('suitable');
		expect(result.riskLevel).toBe('low');
		// Only the positive ASA-I ideal-donor marker fires; it is excluded from penalties.
		expect(result.firedRules.every((r) => r.id === 'SU-001')).toBe(true);
	});

	it('classifies a Grade-2 finding as conditionally-suitable / moderate', () => {
		const d = createIdealLivingDonor();
		d.medicalHistory.hasDiabetes = 'yes';
		const result = gradeDonor(d);
		expect(result.suggestedEligibility).toBe('conditionally-suitable');
		expect(result.riskLevel).toBe('moderate');
	});

	it('classifies a Grade-3 finding as conditionally-suitable / high', () => {
		const d = createIdealLivingDonor();
		d.organFunction.egfr = 45; // OF-002 grade 3
		const result = gradeDonor(d);
		expect(result.suggestedEligibility).toBe('conditionally-suitable');
		expect(result.riskLevel).toBe('high');
	});

	it('classifies a Grade-4 contraindication as unsuitable / critical', () => {
		const d = createIdealLivingDonor();
		d.infectiousDiseaseScreening.hivStatus = 'positive'; // ID-001 grade 4
		const result = gradeDonor(d);
		expect(result.suggestedEligibility).toBe('unsuitable');
		expect(result.riskLevel).toBe('critical');
	});

	it('honours an explicit assessor eligibility decision while preserving risk', () => {
		const d = createIdealLivingDonor();
		d.medicalHistory.hasDiabetes = 'yes'; // moderate risk
		d.eligibilityAllocation.eligibilityDecision = 'unsuitable';
		const result = gradeDonor(d);
		expect(result.eligibility).toBe('unsuitable');
		expect(result.suggestedEligibility).toBe('conditionally-suitable');
		expect(result.riskLevel).toBe('moderate');
	});

	it('has unique rule IDs', () => {
		const ids = donationRules.map((r) => r.id);
		expect(new Set(ids).size).toBe(ids.length);
	});
});

describe('Organ Donation Flagged Issues Detection', () => {
	it('raises no flags for an ideal donor', () => {
		expect(detectAdditionalFlags(createIdealLivingDonor())).toHaveLength(0);
	});

	it('flags a positive HIV screen as high priority', () => {
		const d = createIdealLivingDonor();
		d.infectiousDiseaseScreening.hivStatus = 'positive';
		const flags = detectAdditionalFlags(d);
		expect(flags.some((f) => f.id === 'FLAG-ID-hivStatus' && f.priority === 'high')).toBe(true);
	});

	it('flags coercion concerns for a living donor', () => {
		const d = createIdealLivingDonor();
		d.psychologicalAssessment.coercionConcerns = 'yes';
		const flags = detectAdditionalFlags(d);
		expect(flags.some((f) => f.id === 'FLAG-PS-001')).toBe(true);
	});

	it('does not raise living-donor psychological flags for deceased donors', () => {
		const d = createIdealLivingDonor();
		d.donorTypeRegistration.donorType = 'deceased';
		d.psychologicalAssessment.coercionConcerns = 'yes';
		const flags = detectAdditionalFlags(d);
		expect(flags.some((f) => f.id === 'FLAG-PS-001')).toBe(false);
	});

	it('sorts flags high → medium → low', () => {
		const d = createIdealLivingDonor();
		d.infectiousDiseaseScreening.hivStatus = 'positive'; // high
		d.medicalHistory.hasDiabetes = 'yes'; // medium
		d.medicalHistory.hasHypertension = 'yes'; // low
		const flags = detectAdditionalFlags(d);
		const order = { high: 0, medium: 1, low: 2 };
		const priorities = flags.map((f) => f.priority);
		const sorted = [...priorities].sort((a, b) => order[a] - order[b]);
		expect(priorities).toEqual(sorted);
	});
});
